Platform/Teams/Tools and Improvements/[2019-06-11] State of E2E Testing on VA.gov.md

```diff
@@ -0,0 +1,53 @@
+# State of E2E Testing on VA.gov
+We currently have multiple E2E testing solutions, mostly on the front-end (JavaScript-based). Not every app or feature on VA.gov has full E2E test coverage, as we do not enforce a required amount of coverage to push changes. However, almost all of the existing tests are written in Nightwatch, a testing framework that runs on Selenium, and run as part of CI. These tests do not actually invoke the VA API; when writing these tests, responses need to be mocked for any anticipated API requests.
+
+The Nightwatch tests can run on different browsers through an integration with Sauce Labs, which is a cross-browser testing platform, but we do not run tests on this platform as part of CI. This solution allows us to test on various combinations of OSes and browsers, including IE. However, from cursory attempts at running the full suite of E2E tests across a specific version of each browser under our current (free?) plan, this solution did not appear to provide reliable or timely results. The tests seemed to experience high latency from serving the local assets to the test instance through a proxy, or perhaps loading assets in general. For pages that require a lot of resources such as Facility Locator, the tests frequently timed out. Running a single test on one browser seemed to work fairly consistently but still took a very long time to complete.
+
+There has been some preliminary work done with the Puppeteer testing library. In particular, there is a form filler utility written with Puppeteer that can be extended to work on any of the forms. So far, it has only been applied to an E2E test of the 526 disability claim form. This library is also being leveraged for some visual regression testing.
+
+Finally, there is a WIP integrated testing solution using Capybara for HCA. It runs on Selenium, similar to the Nightwatch tests. The difference is that this will run off of the Rails back-end, with the benefit of being able directly interact with Rails to populate the database with mock data.
+
+
+## Considerations / Priorities
+There are various non-Selenium solutions on the rise with the advantages of running tests faster and being easier to implement. As mentioned, we have some experience already with Puppeteer. Cypress and TestCare are frameworks that we could consider trying out as well.
+
+### Integration with VA API
+The current E2E solution involves creating mock responses within the tests rather than invoking the API. It makes sense to mock responses from third party integrations to limit failures to components that we can control. However, a true E2E test should verify the interaction with the back-end and ensure that the logic there is sound. Note that the API makes use of Betamocks for mock responses, though limited to idempotent actions.
```

Comment:
when you say "the API", is this talking about vets-api (mocking out FE->BE calls) or other downstream backends (behind vets-api)?

Reply:
This refers to `vets-api` mocking downstream back-ends like MVI, MHV, etc.